lib/puppet/functions/simplib/passgen.rb

```diff
@@ -0,0 +1,127 @@
+# vim: set expandtab ts=2 sw=2:
+Puppet::Functions.create_function(:'simplib::passgen') do
+  dispatch :passgen do
+    param          'String', :param
+    optional_param 'Hash',    :options
+  end
+  def gen_random_pass(length,complexity,complex_only)
+
+    length = length.to_i
+    if length.eql?(0)
+      length = @default_password_length
+    elsif length < 8
+      length = 8
+    end
+
+    passwd = ''
+    begin
+      Timeout::timeout(30) do
+        default_charlist = ("a".."z").to_a + ("A".."Z").to_a + ("0".."9").to_a
+        specific_charlist = nil
+        case complexity
+        when 1
+          specific_charlist = ['@','%','-','_','+','=','~']
+        when 2
+          specific_charlist = (' '..'/').to_a + ('['..'`').to_a + ('{'..'~').to_a
+        else
+        end
+        unless specific_charlist == nil
+          if complex_only == true
+            charlists = [
+              specific_charlist
+            ]
+          else
+            charlists = [
+              default_charlist,
+              specific_charlist
+            ]
+          end
+
+        else
+          charlists = [
+            default_charlist
+          ]
+        end
+
+        charlists.each do |charlist|
+          (length/charlists.length).ceil.times { |i|
+            passwd += charlist[rand(charlist.size-1)]
+          }
+        end
+
+        passwd = passwd[0..(length-1)]
```

Comment:
As long as you're here, this might as well be cleaned up.  http://stackoverflow.com/a/88341/2044979

This should probably use `SecureRandom`, as well.